ext/nnstreamer/tensor_filter/tensor_filter_tvm.cc

```diff
@@ -0,0 +1,413 @@
+  gmod = mod_factory.GetFunction ("default") (device);
+
+  if (!set_tensor_info (inputInfo, prop->input_meta, (int) gmod.GetFunction ("get_num_inputs") ()) || 
+      !set_tensor_info (outputInfo, prop->output_meta, (int) gmod.GetFunction ("get_num_outputs") ())) {
+    nns_loge ("Failed to set tensor info.");
+    throw std::invalid_argument ("Failed to set tensor info.");
+  }
+  if (!configure_meta (inputInfo, input_info_list) ||
+      !configure_meta (outputInfo, output_info_list)) {
+    nns_loge ("Failed to configure tensor meta.");
+    throw std::invalid_argument ("Failed to configure tensor meta.");
+  }
+  empty_model = false;
+}
+
+/**
+ * @brief Internal method to set tensor information for model input & output
+ */
+bool tvm_subplugin::configure_meta (GstTensorsInfo &tensor_meta, std::vector<tvm_data_info> &tensor_info_list)
+{
+  for (unsigned int i = 0; i < tensor_meta.num_tensors; ++i) {
+    tvm_data_info tvm_info;
+    auto &info = tensor_meta.info[i];
+    
+    tvm_info.shape = std::vector<int64_t> (std::begin(info.dimension), std::end(info.dimension));
+    if (!convert_nns_type (info.type, tvm_info.dtype)) {
+      nns_loge ("Failed to convert input type.");
+      return false;
+    }
+    tensor_info_list.push_back (tvm_info);
+  }
+  return true;
+}
+
+/**
+ * @brief Internal method to copy user defined tensor properties
+ */
+bool tvm_subplugin::set_tensor_info (GstTensorsInfo &dest_info, const GstTensorsInfo &src_info, unsigned int num_tensors)
+{
+  if (num_tensors > NNS_TENSOR_SIZE_LIMIT)
+  {
+    nns_loge ("The number of tensors required by the given model exceeds the nnstreamer tensor limit (16 by default).");
+    return false;
+  }
+
+  dest_info.num_tensors = num_tensors;
+  for (unsigned int i = 0; i < num_tensors; i++) {
+    gst_tensor_info_copy (std::addressof(dest_info.info[i]), std::addressof(src_info.info[i]));
+  }
+  return true;
+}
+
+/**
+ * @brief Invoke tvm instance
+ */
+void tvm_subplugin::invoke (const GstTensorMemory *input, GstTensorMemory *output) 
+{
+  assert (!empty_model);
+  assert (gmod.defined());
+  assert (input != NULL && output != NULL);
+  unsigned int i;
+  tvm::runtime::NDArray tensor;
+
+  /* read functions */
+  tvm::runtime::PackedFunc set_input = gmod.GetFunction("set_input");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `set_input` not defined in model");
+  tvm::runtime::PackedFunc get_output = gmod.GetFunction("get_output");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `get_output` not defined in model");
+  tvm::runtime::PackedFunc run = gmod.GetFunction("run");
+  if (set_input == nullptr) 
+    throw std::runtime_error ("packed function `run` not defined in model");
+  
+  for (i = 0; i < inputInfo.num_tensors; ++i) {
+    tensor = tvm::runtime::NDArray::Empty(input_info_list[i].shape, input_info_list[i].dtype, device);
+    tensor.CopyFromBytes(input[i].data, input[i].size);
```

Comment:
I wonder if it is required to make a copy?

Reply:
solved. thx